### PR TITLE
perf optimizations for large stream notebook outputs

### PR DESCRIPTION
**Pull request by amunger**
for https://github.com/microsoft/vscode/issues/182636

- skip rendering outputs if we've already sent a message to the renderer for that version of output
- added an 'appended' field to the message object for text that was added since the previous showOutput message (not on the API at the moment)
- streaming output will be appended to a scrollable output element to reduce the amount of DOM manipulation
- adjusted the limits of lines in the scrollable outputs: if a large output comes in, display up to 5000 lines of it (unchanged) but keep appending output up to 8000 lines.
  - once the 8000 line limit is reached, replace it with the last 5000 lines of the full output. That way we can continue to append on top of that

### CODE REVIEW

File: extensions/notebook-renderers/src/textHelper.ts [hunk 1]
```diff
@@ -91,22 +93,46 @@ function truncatedArrayOfString(id: string, buffer: string[], linesLimit: number
...
+	// Allow the output to grow to the hard limit then replace it with the last softLimit number of lines if it grows too large
+	if (buffer.length + outputLengths[id] > hardScrollableLineLimit) {
+		const fullBuffer = fullText.split(/\r\n|\r|\n/g);
+		outputLengths[id] = Math.min(fullBuffer.length, softScrollableLineLimit);
+		const newElement = scrollableArrayOfString(id, fullBuffer.slice(-1 * softScrollableLineLimit), trustHtml);
+		newElement.setAttribute('output-item-id', id);
+		element.replaceWith(newElement);
+	}
+	else {
+		element.appendChild(handleANSIOutput(buffer.join('\n'), trustHtml));
```
Comment (r3m0t) [inline extensions/notebook-renderers/src/textHelper.ts:134]: is this assuming that the previous output always ended with `\n` or `\r` ? It seems only lines are appended to the DOM, characters don't get appended to the existing last line?

File: extensions/notebook-renderers/src/textHelper.ts [hunk 2]
```diff
@@ -91,22 +93,46 @@ function truncatedArrayOfString(id: string, buffer: string[], linesLimit: number
...
+const outputLengths: Record<string, number> = {};
+
+export function appendScrollableOutput(element: HTMLElement, id: string, appended: string, fullText: string, trustHtml: boolean) {
+	if (!outputLengths[id]) {
+		outputLengths[id] = 0;
+	}
+
+	const buffer = appended.split(/\r\n|\r|\n/g);
+	const appendedLength = buffer.length + outputLengths[id];
+	// Allow the output to grow to the hard limit then replace it with the last softLimit number of lines if it grows too large
```
Comment (r3m0t) [inline extensions/notebook-renderers/src/textHelper.ts:125]: `Allow the displayed output to grow to the hard limit then replace it with the last softLimit number of lines if it grows too large`

Initially on reading the code I was worried you reverted to the old behaviour once the line limit was reached, then I realised your `outputLengths` is actually counting the number of lines displayed in the DOM, and the only expensive update is when this exceeds 8000 at which point you reduce back to displaying 5000.

File: extensions/notebook-renderers/src/textHelper.ts (hunk 1 above)
Comment (amunger) [reply extensions/notebook-renderers/src/textHelper.ts:134]: good call on checking that, I confirmed the current implementation will append to the same line

File: extensions/notebook-renderers/src/textHelper.ts [hunk 3]
```diff
@@ -4,9 +4,12 @@
  *--------------------------------------------------------------------------------------------*/
 
 import { handleANSIOutput } from './ansi';
-
+import { OutputElementOptions, OutputWithAppend } from './rendererTypes';
 export const scrollableClass = 'scrollable';
 
+const softScrollableLineLimit = 5000;
+const hardScrollableLineLimit = 8000;
```
Comment (rebornix) [inline extensions/notebook-renderers/src/textHelper.ts:11]: This is very smart

**Review by rebornix — APPROVED** [review]
Great work, I absolutely love it!

**Pull request merged**